File: core/prometheus/loaddashboards.go
package prometheus

import (
	"encoding/json"
	"flag"
	"io/ioutil"
	"log"
	"net/http"

	"gopkg.in/yaml.v2"
)

type Template struct {
	Name  string
	Query string
	Value string
}

type Dashboard struct {
	Name          string
	TargetQueries []string
	Annoation     string
	Templates     []Template
	Period        string
	Interval      string
}

type dashboardsDefenition struct {
	Items []struct {
		Data map[string]string `yaml:"data"`
	} `yaml:"items"`
}

type jsonDahboard struct {
	Rows []struct {
		Panels []struct {
			Targets []struct {
				Expr string `json:"expr"`
			} `json: "targets"`
		} `json: "panels"`
	} `json: "rows"`
	Templating struct {
		List []struct {
			Current struct {
				Value string `json: "value"`
			} `json: "current"`
			Name  string `json: "name"`
			Query string `json: "query"`
		} `json: "list"`
	} `json: "templating"`
	Time struct {
		From string `json: "from"`
	} `json: "time"`
	Refresh string `json: "refresh"`
}

func (yml *dashboardsDefenition) getyaml(yamlfile string) *dashboardsDefenition {
	var client http.Client
	resp, err := client.Get(yamlfile)
	if err != nil {
		// err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		bodyBytes, _ := ioutil.ReadAll(resp.Body)
		err = yaml.Unmarshal(bodyBytes, yml)
		if err != nil {
			log.Fatalf("Unmarshal: %v", err)
		}
	}
	return yml
}

func (jd *jsonDahboard) GetRefreshInterval() string {
	return jd.Refresh
}

func (jd *jsonDahboard) GetTimePeriod() string {
	return jd.Time.From
}

func (jd *jsonDahboard) GetTemplates() []Template {
	templates := make([]Template, len(jd.Templating.List))
	for _, template := range jd.Templating.List {
		tmp := Template{Name: template.Name, Query: template.Query, Value: template.Current.Value}
		templates = append(templates, tmp)
	}
	return templates
}

func (jd *jsonDahboard) GetTargets() []string {
	targets := make([]string, 0)
	for _, row := range jd.Rows {
		for _, panel := range row.Panels {
			for _, target := range panel.Targets {
				targets = append(targets, target.Expr)
			}
		}
	}
	return targets
}

func DashboardLoader() []Dashboard {
	var yaml string
	flag.StringVar(&yaml, "yaml", "https://raw.githubusercontent.com/openshift/cluster-monitoring-operator/master/assets/grafana/dashboard-definitions.yaml",
		"link to a dashboard definitation file")
	flag.Parse()

	var yml dashboardsDefenition
	yml.getyaml(yaml)

	dashboards := make([]Dashboard, 0, len(yml.Items))
	var jsondahboard jsonDahboard
	for _, dashboard := range yml.Items {
		for name, jsondata := range dashboard.Data {
			json.Unmarshal([]byte(jsondata), &jsondahboard)
			dashboards = append(dashboards, Dashboard{
				Name:          name,
				TargetQueries: jsondahboard.GetTargets(),
				Templates:     jsondahboard.GetTemplates(),
				Period:        jsondahboard.GetTimePeriod(),
				Interval:      jsondahboard.GetRefreshInterval(),
				Annoation:     "",
			})
		}
	}
	return dashboards
}

File: core/prometheus/loaddashboards.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""
The following module will load and parse the dashboard-definitions from
cluster-monitoring-operator project and retrieve dashboards object that include
a set of dashboard and queries dictionaries.
"""
import yaml
import json
import os
import argparse
from requests.utils import quote


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('-g',
                        '--gitrepo',
                        type=str,
                        required=False,
                        default='https://raw.githubusercontent.com/openshift/cluster-monitoring-operator/master/assets/grafana/dashboard-definitions.yaml',
                        dest='gitrepo',
                        help='gitrepo')
    parser.add_argument('-p',
                        '--print',
                        type=str,
                        required=False,
                        dest='printit',
                        help='print results')
    return parser.parse_args()

class Dashboards(object):
    def __init__(self, file):
        self.dashboards = []

        if 'http' in file:
            import urllib
            sock = urllib.urlopen(file)
            htmlSource = sock.read()
            sock.close()
            self.yaml = yaml.load(htmlSource)
        else:
            self.yaml = self.load_yaml(file)

        self.scan_dashboards()

    def load_yaml(self, f):
        with open(f, 'r') as y:
            return yaml.load(y)

    def scan_dashboards(self):
        for item in self.yaml.get('items'):
            dashboard = {}
            dashboard['name'] = item['data'].keys()[0]
            dashboard['queries'] = self.scan_queries(json.loads(item['data'].values()[0]))
            self.dashboards.append(dashboard)

    def scan_queries(self, jsondata):
        exprs = []
        for raw in jsondata['rows']:
            for panel in raw['panels']:
                for target in panel['targets']:
                    exprs.append(quote(target['expr']))
        return exprs

    def get_dashboards(self):
        return self.dashboards

if __name__ == "__main__":
    args = parse_args()
    loader = Dashboards(args.gitrepo)
    loader.get_dashboards()
